pkg/controller/proxyconfig/controller.go

```diff
@@ -0,0 +1,147 @@
+	}
+
+	return &ReconcileProxyConfig{client: mgr.GetClient(), scheme: mgr.GetScheme(), status: status}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("proxyconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the proxy resource.
+	err = c.Watch(&source.Kind{Type: &configv1.Proxy{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ReconcileProxyConfig reconciles a Proxy object
+type ReconcileProxyConfig struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver.
+	client client.Client
+	scheme *runtime.Scheme
+	status *statusmanager.StatusManager
+}
+
+// Reconcile expects request to refer to a proxy object named "cluster"
+// in the default namespace and will ensure the proxy object is in the
+// desired state.
+func (r *ReconcileProxyConfig) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Printf("Reconciling proxy: %s\n", request.Name)
+	proxyConfig := &configv1.Proxy{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, proxyConfig)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue
+			log.Println("proxy not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("failed to get proxy %q: %v", request, err)
+	}
+
+	// A nil proxy is generated by upgrades and installs not requiring a proxy.
+	if !isSpecHTTPProxySet(&proxyConfig.Spec) &&
+		!isSpecHTTPSProxySet(&proxyConfig.Spec) &&
+		!isSpecNoProxySet(&proxyConfig.Spec) {
+		log.Printf("httpProxy, httpsProxy and noProxy not defined; validation will be skipped for proxy: %s\n",
+			request.Name)
+	}
+
+	// Only proceed if the required config objects can be collected.
+	infraConfig := &configv1.Infrastructure{}
+	netConfig := &configv1.Network{}
+	clusterConfig := &corev1.ConfigMap{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, infraConfig); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to get infrastructure config 'cluster': %v", err)
+	}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, netConfig); err != nil {
+		log.Printf("failed to get network config 'cluster': %v", err)
+		return reconcile.Result{}, err
+	}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster-config-v1", Namespace: "kube-system"},
+		clusterConfig); err != nil {
+		log.Printf("failed to get configmap 'kube-system/cluster-config-v1': %v", err)
+		return reconcile.Result{}, err
+	}
```

Comment:
should these 3 failure cases set the degraded status also?  they all mean we effectively ignored the proxy spec

Reply:
@bparees yes, let me update the PR. As with `r.ValidateProxyConfig()` returning `err != nil`, I'm also going to update each call to `return reconcile.Result{}, nil` upon `err != nil`.